File: 2022/Day 04/pt2.cpp
#include "helper.hpp"

int get_partial_overlaps(std::vector<std::string> lines)
{
    int partial_overlaps = 0;

    for (std::string line : lines)
    {
        int l0, l1, r0, r1;
        sscanf(line.c_str(), "%d-%d,%d-%d", &l0, &l1, &r0, &r1);
        partial_overlaps += ((l0 <= r0 && l1 >= r0) || (r0 <= l0 && r1 >= l0));
    }

    return partial_overlaps;
}

int main()
{
    std::vector<std::string> lines = Helper::get_lines("input.txt");

    auto t1 = Helper::get_current_time();
    int partial_overlaps = get_partial_overlaps(lines);
    std::cout << "Partial Overlaps: " << partial_overlaps << std::endl;
    auto t2 = Helper::get_current_time();
    Helper::get_time_taken(t1, t2);

    return 0;
}

File: 2023/day_02/pt2.go
package day02

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"

	"github.com/benjababe/advent-of-code/helper"
)

func solve(lines []string) int {
	score := 0

	for _, line := range lines {
		line = line[strings.Index(line, ":")+1:]
		minCubes := map[string]int{"r": 0, "g": 0, "b": 0}

		r := regexp.MustCompile(`(\d+)\s(r|g|b)`)
		matches := r.FindAllStringSubmatch(line, -1)

		for _, match := range matches {
			count, _ := strconv.Atoi(match[1])
			if count > minCubes[match[2]] {
				minCubes[match[2]] = count
			}
		}

		score += minCubes["r"] * minCubes["g"] * minCubes["b"]
	}

	return score
}

func Pt2() {
	lines := []string{}
	helper.GetLines(&lines, "input.txt")

	start := helper.GetCurrentTime()
	output := solve(lines)
	fmt.Printf("Day 2\tPt2:\t%d\n", output)
	end := helper.GetCurrentTime()
	helper.GetTimeTaken(start, end)

}
